ADApp/pluginSrc/NDFileHDF5AttributeDataset.cpp

```diff
@@ -185,36 +185,61 @@ asynStatus NDFileHDF5AttributeDataset::writeAttributeDataset(hdf5::When_t whenTo
       extendIndexDataSet(offsets[indexed]);
     }
     // find the data based on datatype
-    ret = ndAttr->getValue(ndAttr->getDataType(), pDatavalue, MAX_ATTRIBUTE_STRING_SIZE);
+    ret = ndAttr->getValue(ndAttr->getDataType(), pDatavaluestore[dataset_count], MAX_ATTRIBUTE_STRING_SIZE);
     if (ret == ND_ERROR) {
-      memset(pDatavalue, 0, MAX_ATTRIBUTE_STRING_SIZE);
+      memset(pDatavaluestore[dataset_count], 0, MAX_ATTRIBUTE_STRING_SIZE);
+    }
+    dataset_count++;
+    // Check if we are being asked to flush or if the size of the store has reached its maximum amount
+    if (flush == 1 or dataset_count == MAX_BATCH_SIZE){
+      elementSize_[0] = dataset_count;
+      // Offset has been incremented, so is at where the stored data should end
+      offset_[0] -= (dataset_count - 1);
+      // Work with HDF5 library to select a suitable hyperslab (one element) and write the new data to it
+      H5Dset_extent(dataset_, dims_);
+      filespace_ = H5Dget_space(dataset_);
+      memspace_ = H5Screate_simple(rank_, elementSize_, NULL);
+      // Select the hyperslab
+      H5Sselect_hyperslab(filespace_, H5S_SELECT_SET, offset_, NULL, elementSize_, NULL);
+      H5Dwrite(dataset_, datatype_, memspace_, filespace_, H5P_DEFAULT, pDatavaluestore);
+      if (flush ==1)
+        status = this->flushDataset();
+      H5Sclose(filespace_);
+      dataset_count = 0;
     }
+
+    nextRecord_++;
+  }
+
+  return status;
+}
```

Comment:
This raises another issue that is not directly related to these changes, which is that the write H5 calls in the two (now maybe three) similar methods are duplicated and functionally the same. It looks like one version has had an error check about the data being undefined added, which should probably be done in both. Is it worth consolidating this, especially if we do want to add a third instance of what is almost the same thing?